borsh/src/de/mod.rs

```diff
@@ -472,37 +475,119 @@ pub mod hashes {
     use crate::BorshDeserialize;
     use crate::__maybestd::collections::{HashMap, HashSet};
     use crate::__maybestd::io::{Read, Result};
-    use crate::__maybestd::vec::Vec;
+    use crate::de::hint;
+
+    #[cfg(feature = "de_strict_order")]
+    const ERROR_WRONG_ORDER_OF_KEYS: &str = "keys were not serialized in ascending order";
+    #[cfg(feature = "de_strict_order")]
+    use crate::__maybestd::io::{Error, ErrorKind};
+    #[cfg(feature = "de_strict_order")]
+    use core::cmp::Ordering;
 
     impl<T, H> BorshDeserialize for HashSet<T, H>
     where
-        T: BorshDeserialize + Eq + Hash,
+        T: BorshDeserialize + Eq + Hash + PartialOrd,
         H: BuildHasher + Default,
     {
         #[inline]
         fn deserialize_reader<R: Read>(reader: &mut R) -> Result<Self> {
-            let vec = <Vec<T>>::deserialize_reader(reader)?;
-            Ok(vec.into_iter().collect::<HashSet<T, H>>())
+            let len = u32::deserialize_reader(reader)?;
```

Comment:
- [x] `HashSet/BTreeSet/HashMap/BTreeMap` impl
    - [x] `HashSet` 
      - [x] rewrite to no extra `Vec` allocation
      - [x] `#cfg(...)` addon
    - [x] `HashMap` 
      - [x] add `length_hint` in capacity allocation
      - [x] `#cfg(...)` addon
    - [x] `BTreeSet` 
      - [x] leave it as is, as `BTreeSet` 
        - [x] has optimisation for initialization from Iterator (with sorting)
          - [x] [set source](https://doc.rust-lang.org/src/alloc/collections/btree/set.rs.html#1210)
      - [x] `#cfg(...)` addon
    - [x] `BTreeMap`
      - [x] replace with parsing to `Vec` first
        - [x] as `BTreeMap` has optimisation for init from Iterator (with sorting) 
          - [x] [map source](https://doc.rust-lang.org/src/alloc/collections/btree/map.rs.html#2237)
      - [x] `#cfg(...)` addon

Reply:
@dj8yfo Let's mention those optimizations for `from_iter` in the source code, so whoever will look into it in the future will immediately know that